// @cornerstonejs/tools/dist/esm/stateManagement/segmentation/polySeg/Contour/contourComputationStrategies.js
import { cache } from '@cornerstonejs/core';
import { getUniqueSegmentIndices } from '../../../../utilities/segmentation';
import { getSegmentation, setSegmentationRepresentationSpecificConfig, } from '../../segmentationState';
import { computeSurfaceFromLabelmapSegmentation } from '../Surface/surfaceComputationStrategies';
import { clipAndCacheSurfacesForViewport, } from '../../helpers/clipAndCacheSurfacesForViewport';
import { extractContourData } from './utils/extractContourData';
import { createAndAddContourSegmentationsFromClippedSurfaces } from './utils/createAndAddContourSegmentationsFromClippedSurfaces';
import { getToolGroupForViewport } from '../../../../store/ToolGroupManager';
export async function computeContourData(segmentationId, options = {}) {
    const segmentIndices = options.segmentIndices?.length
        ? options.segmentIndices
        : getUniqueSegmentIndices(segmentationId);
    let rawContourData;
    const segmentation = getSegmentation(segmentationId);
    const representationData = segmentation.representationData;
    try {
        if (representationData.SURFACE) {
            rawContourData = await computeContourFromSurfaceSegmentation(segmentationId, {
                segmentIndices,
                ...options,
            });
        }
        else if (representationData.LABELMAP) {
            rawContourData = await computeContourFromLabelmapSegmentation(segmentationId, {
                segmentIndices,
                ...options,
            });
        }
    }
    catch (error) {
        console.error(error);
        throw error;
    }
    if (!rawContourData) {
        throw new Error('Not enough data to convert to contour, currently only support converting volume labelmap to contour if available');
    }
    const { viewport, segmentationRepresentationUID } = options;
    const annotationUIDsMap = createAndAddContourSegmentationsFromClippedSurfaces(rawContourData, viewport, segmentationId);
    const toolGroupId = getToolGroupForViewport(viewport.id)?.id;
    setSegmentationRepresentationSpecificConfig(toolGroupId, segmentationRepresentationUID, {
        CONTOUR: {
            fillAlpha: 0,
        },
    });
    return {
        annotationUIDsMap,
    };
}
async function computeContourFromLabelmapSegmentation(segmentationId, options = {}) {
    if (!options.viewport) {
        throw new Error('Viewport is required to compute contour from labelmap');
    }
    const results = await computeSurfaceFromLabelmapSegmentation(segmentationId, options);
    if (!results?.length) {
        console.error('Failed to convert labelmap to surface or labelmap is empty');
        return;
    }
    const { viewport, segmentationRepresentationUID } = options;
    const pointsAndPolys = results.map((surface) => {
        return {
            id: surface.segmentIndex.toString(),
            points: surface.data.points,
            polys: surface.data.polys,
            segmentIndex: surface.segmentIndex,
        };
    });
    const polyDataCache = await clipAndCacheSurfacesForViewport(pointsAndPolys, viewport, segmentationRepresentationUID);
    const rawResults = extractContourData(polyDataCache);
    return rawResults;
}
async function computeContourFromSurfaceSegmentation(segmentationId, options = {}) {
    if (!options.viewport) {
        throw new Error('Viewport is required to compute contour from surface');
    }
    const { viewport, segmentationRepresentationUID } = options;
    const segmentIndices = options.segmentIndices?.length
        ? options.segmentIndices
        : getUniqueSegmentIndices(segmentationId);
    const segmentIndexToSurfaceId = new Map();
    const surfaceIdToSegmentIndex = new Map();
    const segmentation = getSegmentation(segmentationId);
    const representationData = segmentation.representationData.SURFACE;
    const surfacesInfo = [];
    representationData.geometryIds.forEach((geometryId, segmentIndex) => {
        if (segmentIndices.includes(segmentIndex)) {
            segmentIndexToSurfaceId.set(segmentIndex, geometryId);
            const surface = cache.getGeometry(geometryId)?.data;
            if (surface) {
                surfacesInfo.push({
                    id: geometryId,
                    points: surface.getPoints(),
                    polys: surface.getPolys(),
                });
            }
        }
    });
    segmentIndexToSurfaceId.forEach((surfaceId, segmentIndex) => {
        surfaceIdToSegmentIndex.set(surfaceId, segmentIndex);
    });
    const polyDataCache = await clipAndCacheSurfacesForViewport(surfacesInfo, viewport, segmentationRepresentationUID);
    const rawResults = extractContourData(polyDataCache, surfaceIdToSegmentIndex);
    return rawResults;
}
export { computeContourFromLabelmapSegmentation };
