// @cornerstonejs/tools/dist/esm/stateManagement/segmentation/helpers/clipAndCacheSurfacesForViewport.js
import { Enums, getWebWorkerManager, eventTarget, triggerEvent, } from '@cornerstonejs/core';
import { WorkerTypes } from '../../../enums';
import { pointToString } from '../../../utilities';
import { registerPolySegWorker } from '../polySeg/registerPolySegWorker';
const workerManager = getWebWorkerManager();
const polyDataCache = new Map();
const surfacesAABBCache = new Map();
const triggerWorkerProgress = (eventTarget, progress) => {
    triggerEvent(eventTarget, Enums.Events.WEB_WORKER_PROGRESS, {
        progress,
        type: WorkerTypes.SURFACE_CLIPPING,
    });
};
export async function clipAndCacheSurfacesForViewport(surfacesInfo, viewport, segmentationRepresentationUID) {
    registerPolySegWorker();
    const planesInfo = viewport.getSlicesClippingPlanes?.();
    if (!planesInfo) {
        return;
    }
    const currentSliceIndex = viewport.getSliceIndex();
    planesInfo.sort((a, b) => {
        const diffA = Math.abs(a.sliceIndex - currentSliceIndex);
        const diffB = Math.abs(b.sliceIndex - currentSliceIndex);
        return diffA - diffB;
    });
    triggerWorkerProgress(eventTarget, 0);
    await updateSurfacesAABBCache(surfacesInfo);
    const surfacesAABB = new Map();
    surfacesInfo.forEach((surface) => {
        surfacesAABB.set(surface.id, surfacesAABBCache.get(surface.id));
    });
    const camera = viewport.getCamera();
    await workerManager
        .executeTask('polySeg', 'cutSurfacesIntoPlanes', {
        surfacesInfo,
        planesInfo,
        surfacesAABB,
    }, {
        callbacks: [
            ({ progress }) => {
                triggerWorkerProgress(eventTarget, progress);
            },
            ({ sliceIndex, polyDataResults }) => {
                polyDataResults.forEach((polyDataResult, surfaceId) => {
                    const actorUID = `${segmentationRepresentationUID}_${surfaceId}`;
                    const cacheId = generateCacheId(viewport, camera.viewPlaneNormal, sliceIndex);
                    updatePolyDataCache(actorUID, cacheId, polyDataResult);
                });
            },
        ],
    })
        .catch((error) => {
        console.error(error);
    });
    triggerWorkerProgress(eventTarget, 1);
    return polyDataCache;
}
async function updateSurfacesAABBCache(surfacesInfo) {
    const surfacesWithoutAABB = surfacesInfo.filter((surface) => !surfacesAABBCache.has(surface.id));
    if (!surfacesWithoutAABB.length) {
        return;
    }
    const surfacesAABB = await workerManager.executeTask('polySeg', 'getSurfacesAABBs', {
        surfacesInfo: surfacesWithoutAABB,
    }, {
        callbacks: [
            ({ progress }) => {
                triggerWorkerProgress(eventTarget, progress);
            },
        ],
    });
    surfacesAABB.forEach((aabb, id) => {
        surfacesAABBCache.set(id, aabb);
    });
}
export function getSurfaceActorUID(segmentationRepresentationUID, surfaceId) {
    return `${segmentationRepresentationUID}_${surfaceId}`;
}
export function generateCacheId(viewport, viewPlaneNormal, sliceIndex) {
    return `${viewport.id}-${pointToString(viewPlaneNormal)}-${sliceIndex}`;
}
export function updatePolyDataCache(actorUID, cacheId, polyDataResult) {
    const { points, lines, numberOfCells } = polyDataResult;
    let actorCache = polyDataCache.get(actorUID);
    if (!actorCache) {
        actorCache = new Map();
        polyDataCache.set(actorUID, actorCache);
    }
    actorCache.set(cacheId, { points, lines, numberOfCells });
}
